# Contents

01. Importing Libraries
02. Importing Data
03. Investigate the Data Frame
04. Cleaning the Data
05. Consider security implications exist in this data. (PII Data)
06. Comparing Customer Behaviour in Different Geographic Areas
07. Create exclusion flag for low-activity customer (less than 5)

# 01. Importing Libraries

In [16]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02. Importing Data

In [17]:
# Import the data set into Jupyter as a new dataframe
path = r'/Users/ChuahLH/Desktop/Documents/Instacart Basket Analysis'

In [18]:
# Import the final merged dataframe
df_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_merged_final.pkl'))

# 03. Investigate the Data Frame

In [19]:
df_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  ...         Frequency_flag first_name  last_name  gender  \
0        0.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
1        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
2        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
3        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
4        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   

     state age date_joined number_of_dependants family_status income  
0  Alabama  31   2/17/2019                    3       married  40423  
1  Alabama  31   2/17/2019                    3       married  40423  
2  Alabama  31   2/17/2019                    3       married  40423  
3  Alabama  31   2/17/2019                    3       married  40423  
4  Alabama  31   2/17/2019                    3       married  40423  

[5 rows x 36 columns]

In [20]:
df_merged.shape

(32406041, 36)

# 04. Cleaning the Data

In [21]:
# List all columns info
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32406041 entries, 0 to 32406040
Data columns (total 36 columns):
 #   Column                        Dtype   
---  ------                        -----   
 0   Unnamed: 0_x                  int64   
 1   order_id                      int64   
 2   user_id                       int64   
 3   order_number                  int64   
 4   orders_day_of_week            int64   
 5   order_hour_of_day             int64   
 6   days_since_last_order         float64 
 7   product_id                    float64 
 8   add_to_cart_order             float64 
 9   reordered                     float64 
 10  Unnamed: 0_y                  int64   
 11  product_name                  object  
 12  aisle_id                      int64   
 13  department_id                 int64   
 14  prices                        float64 
 15  _merge                        category
 16  price_range_loc               object  
 17  busiest_day                   object  
 18  

In [22]:
#Renaming the product_id and Unnamed: 0_y
df_merged.rename(columns={'Unnamed: 0_y':'product_id'}, inplace=True)

In [23]:
# Drop the Unnamed: 0_x as it is same as order_id; _merge could be dropped to save the space
df_merged_2 = df_merged.drop(columns=['Unnamed: 0_x','_merge']) 

In [24]:
df_merged_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

   product_id  ...         Frequency_flag  first_name  last_name  gender  \
0         195  ...  Non-Frequent customer       Linda     Nguyen  Female   
1         195  ...  Non-Frequent customer       Linda     Nguyen  Female   
2         195  ...  Non-Frequent customer       Linda     Nguyen  Female   
3         195  ...  Non-Frequent customer       Linda     Nguyen  Female   
4         195  ...  Non-Frequent customer       Linda     Nguyen  Female   

     state age date_joined number_of_dependants  family_status income  
0  Alabama  31   2/17/2019                    3        married  40423  
1  Alabama  31   2/17/2019                    3        married  40423  
2  Alabama  31   2/17/2019                    3        married  40423  
3  Alabama  31   2/17/2019                    3        married  40423  
4  Alabama  31   2/17/2019                    3        married  40423  

[5 rows x 34 columns]

In [25]:
# Check columns to ensure unneccessary columns successfully dropped
df_merged_2.shape

(32406041, 34)

# 05. Consider security implications exist in this data. (PII Data)

Column first_name and last_name clearly contain PII data and should be removed. 

In [26]:
# Drop the first_name and last_name
df_merged_final = df_merged_2.drop(columns=['first_name','last_name']) 

In [27]:
df_merged_final.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

   product_id  ... median_days_since_last_order          frequent_flag  \
0         195  ...                    20.259259  Non-frequent customer   
1         195  ...                    20.259259  Non-frequent customer   
2         195  ...                    20.259259  Non-frequent customer   
3         195  ...                    20.259259  Non-frequent customer   
4         195  ...                    20.259259  Non-frequent customer   

          Frequency_flag  gender    state age date_joined  \
0  Non-Frequent customer  Female  Alabama  31   2/17/2019   
1  Non-Frequent customer  Female  Alabama  31   2/17/2019   
2  Non-Frequent customer  Female  Alabama  31   2/17/2019   
3  Non-Frequent customer  Female  Alabama  31   2/17/2019   
4  Non-Frequent customer  Female  Alabama  31   2/17/2019   

  number_of_dependants  family_status income  
0                    3        married  40423  
1                    3        married  40423  
2                    3        married  40423  
3                    3        married  40423  
4                    3        married  40423  

[5 rows x 32 columns]

In [28]:
df_merged_final.shape

(32406041, 32)

# 06. Comparing Customer Behaviour in Different Geographic Areas

In [29]:
# Use the region information in Wikipedia article to create your column
northeast = ['Maine', 'New Hampshire', 'Vermont', 'Massachusetts', 'Rhode Island', 
             'Connecticut', 'New York', 'Pennsylvania', 'New Jersey']
midwest = ['Wisconsin', 'Michigan', 'Illinois', 'Indiana', 'Ohio',
          'North Dakota', 'South Dakota', 'Nebraska', 'Kansas',
          'Minnesota', 'Iowa', 'Missouri']
south = ['Delaware', 'Maryland', 'District of Columbia', 'Virginia', 'West Virginia',
         'North Carolina', 'South Carolina', 'Georgia', 'Florida', 'Kentucky', 'Tennessee',
         'Mississippi', 'Alabama', 'Oklahoma', 'Texas', 'Arkansas', 'Louisiana']
west = ['Idaho', 'Montana', 'Wyoming', 'Nevada', 'Utah', 'Colorado', 'Arizona',
        'New Mexico', 'Alaska', 'Washington', 'Oregon', 'California', 'Hawaii']

In [30]:
#Creating regional flags
df_merged_final.loc[df_merged_final['state'].isin(northeast), 'region']='northeast'
df_merged_final.loc[df_merged_final['state'].isin(midwest), 'region']='midwest'
df_merged_final.loc[df_merged_final['state'].isin(south), 'region']='south'
df_merged_final.loc[df_merged_final['state'].isin(west), 'region']='west'

In [31]:
#check
df_merged_final['region'].value_counts(dropna=False)

south        10792300
west          8293217
midwest       7597585
northeast     5722939
Name: region, dtype: int64

In [32]:
# Check the region count = df_merged_final
10792300 + 8293217 + 7597585 + 5722939

32406041

In [33]:
df_merged_final.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

   product_id  ...          frequent_flag         Frequency_flag  gender  \
0         195  ...  Non-frequent customer  Non-Frequent customer  Female   
1         195  ...  Non-frequent customer  Non-Frequent customer  Female   
2         195  ...  Non-frequent customer  Non-Frequent customer  Female   
3         195  ...  Non-frequent customer  Non-Frequent customer  Female   
4         195  ...  Non-frequent customer  Non-Frequent customer  Female   

     state age date_joined number_of_dependants family_status  income region  
0  Alabama  31   2/17/2019                    3       married   40423  south  
1  Alabama  31   2/17/2019                    3       married   40423  south  
2  Alabama  31   2/17/2019                    3       married   40423  south  
3  Alabama  31   2/17/2019                    3       married   40423  south  
4  Alabama  31   2/17/2019                    3       married   40423  south  

[5 rows x 33 columns]

In [34]:
df_merged_final.shape

(32406041, 33)

In [35]:
# Check the spending habits between the different U.S. Regions
crosstab=pd.crosstab(df_merged_final['region'], df_merged_final['spending_flag'],  dropna=False)
crosstab

spending_flag  High spender  Low spender
region                                  
midwest              155979      7441606
northeast            108226      5614713
south                209693     10582607
west                 160354      8132863

The high spender region was in south (209,693), then followed by west (160,354), midwest (155,979) and northeast (108,226). Interesting to see the low spender was also in south. It could be due to the customers population in south region, number of states in south region. Investigate the customers population in each region and number of states in each region. 

In [36]:
# Performing aggregation
df_merged_final.groupby('region').agg({'prices': ['mean', 'min', 'max']})

prices              
                mean  min      max
region                            
midwest    12.715281  1.0  99999.0
northeast  11.554842  1.0  99999.0
south      12.195373  1.0  99999.0
west       11.319669  1.0  99999.0

Max prices of 99,999 appeared in the data frame, but the mean is between 11-12. 

In [37]:
df_merged_final['prices'].max()

99999.0

In [38]:
df_merged_final.loc[df_merged_final['prices'] > 100]

order_id  user_id  order_number  orders_day_of_week  \
7805       1435153     1519            26                   2   
7806       1066435     1519            32                   3   
15143      1697208     2409            33                   1   
20264       965345     3531            27                   2   
20497      2943740     3793             1                   2   
...            ...      ...           ...                 ...   
32304980   3265389    51346             4                   1   
32309063   1179092   131671            12                   4   
32311991   1226705    39667             7                   2   
32331229   3000037    95105             2                   4   
32331230   2741225    95105             4                   5   

          order_hour_of_day  days_since_last_order  product_id  \
7805                     11                    6.0     21553.0   
7806                     10                    5.0     21553.0   
15143                    19                    5.0     21553.0   
20264                    19                    8.0     33664.0   
20497                     9                    NaN     21553.0   
...                     ...                    ...         ...   
32304980                 17                   13.0     33664.0   
32309063                  9                   15.0     33664.0   
32311991                 15                   20.0     33664.0   
32331229                 19                   22.0     33664.0   
32331230                 12                    2.0     33664.0   

          add_to_cart_order  reordered  product_id  ...  \
7805                   15.0        0.0       21554  ...   
7806                   28.0        1.0       21554  ...   
15143                   1.0        0.0       21554  ...   
20264                   2.0        0.0       33666  ...   
20497                   3.0        0.0       21554  ...   
...                     ...        ...         ...  ...   
32304980                2.0        1.0       33666  ...   
32309063                1.0        0.0       33666  ...   
32311991               14.0        0.0       33666  ...   
32331229                4.0        0.0       33666  ...   
32331230                3.0        1.0       33666  ...   

                  frequent_flag         Frequency_flag  gender  \
7805                        NaN      Frequent customer    Male   
7806                        NaN      Frequent customer    Male   
15143                       NaN      Frequent customer  Female   
20264                       NaN       Regular customer  Female   
20497                       NaN       Regular customer    Male   
...                         ...                    ...     ...   
32304980                    NaN       Regular customer    Male   
32309063                    NaN       Regular customer  Female   
32311991  Non-frequent customer  Non-Frequent customer    Male   
32331229                    NaN      Frequent customer  Female   
32331230                    NaN      Frequent customer  Female   

                   state age date_joined number_of_dependants  \
7805        Rhode Island  68    3/4/2020                    0   
7806        Rhode Island  68    3/4/2020                    0   
15143             Hawaii  55  11/11/2018                    2   
20264             Hawaii  80   10/6/2019                    2   
20497          Louisiana  59   8/23/2019                    3   
...                  ...  ..         ...                  ...   
32304980    Rhode Island  58   1/25/2019                    1   
32309063    Rhode Island  35    8/4/2018                    0   
32311991    Rhode Island  78    5/8/2019                    2   
32331229  South Carolina  48   1/18/2017                    2   
32331230  South Carolina  48   1/18/2017                    2   

             family_status  income     region  
7805      divorced/widowed  177864  northeast  
7806      divorced/widowed  177864  northeast  
15143    

In [39]:
df_merged_final.loc[df_merged_final['prices'] > 100, 'prices'] = np.nan

In [40]:
df_merged_final['prices'].max()

25.0

In [41]:
df_merged_final.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
..       ...      ...           ...                 ...                ...   
95   3317979       15             5                   4                 15   
96   2685110       15             7                   1                 11   
97    887727       15             9                   2                 13   
98   2600170       15            11                   2                  9   
99   2332087       15            13                   1                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
..                    ...         ...                ...        ...   
95                   17.0     14715.0                1.0        1.0   
96                   17.0     14715.0                3.0        1.0   
97                    7.0     14715.0                1.0        1.0   
98                   14.0     14715.0                1.0        1.0   
99                   14.0     14715.0                3.0        1.0   

    product_id  ...          frequent_flag         Frequency_flag  gender  \
0          195  ...  Non-frequent customer  Non-Frequent customer  Female   
1          195  ...  Non-frequent customer  Non-Frequent customer  Female   
2          195  ...  Non-frequent customer  Non-Frequent customer  Female   
3          195  ...  Non-frequent customer  Non-Frequent customer  Female   
4          195  ...  Non-frequent customer  Non-Frequent customer  Female   
..         ...  ...                    ...                    ...     ...   
95       14715  ...                    NaN       Regular customer  Female   
96       14715  ...                    NaN       Regular customer  Female   
97       14715  ...                    NaN       Regular customer  Female   
98       14715  ...                    NaN       Regular customer  Female   
99       14715  ...                    NaN       Regular customer  Female   

      state age date_joined number_of_dependants     family_status  income  \
0   Alabama  31   2/17/2019                    3           married   40423   
1   Alabama  31   2/17/2019                    3           married   40423   
2   Alabama  31   2/17/2019                    3           married   40423   
3   Alabama  31   2/17/2019                    3           married   40423   
4   Alabama  31   2/17/2019                    3           married   40423   
..      ...  ..         ...                  ...               ...     ...   
95  Indiana  69    6/3/2019                    0  divorced/widowed   54313   
96  Indiana  69    6/3/2019                    0  divorced/widowed   54313   
97  Indiana  69    6/3/2019                    0  divorced/widowed   54313   
98  Indiana  69    6/3/2019                    0  divorced/widowed   54313   
99  Indiana  69    6/3/2019                    0  divorced/widowed   54313   

     region  
0     south  
1     south  
2     south  
3     south  
4     south  
..      ...  
95  midwest  
96  midwest  
97  midwest  
98  midwest  
99  midwest  

[100 rows x 33 columns]

In [42]:
df_merged_final.shape

(32406041, 33)

In [43]:
# Performing aggregation
df_merged_final.groupby('region').agg({'prices': ['mean', 'min', 'max']})

prices           
               mean  min   max
region                        
midwest    7.792550  1.0  25.0
northeast  7.782387  1.0  25.0
south      7.794883  1.0  25.0
west       7.790320  1.0  25.0

The price of the product is approximately same at all regions.

In [44]:
# Performing aggregation
df_merged_final.groupby('region')['state'].nunique()

region
midwest      12
northeast     9
south        17
west         13
Name: state, dtype: int64

There are more states in region south, followed by west, midwest then northeast. 

In summary, there are more regions in south (17) of U.S. It makes sense that the largest customer population is also in south (10,792,300). Due to the largest number of customer population in south, the data frame presented the high spender in the south (209,693) and the lower spender also in south (10,582,607). The average price of the product is about the same across the regions. 

The spending behaviour for each region is about the same; the highest spender can be found in the large population of region; the lowest customer population, the spender is less. 

# 07. Create exclusion flag for low-activity customer (less than 5)

In [45]:
#creating variable for number of orders made by user, each order is represented by entry with user_id in table
#count of user_id would be number of orders
df_merged_final['total_orders']=df_merged_final.groupby(['user_id']).user_id.transform('count')

In [46]:
#Creating exclusion flag for low-activity customers: less than 5 and equal to or more than 5
results_2 = []
for total in df_merged_final['total_orders']:
    if total < 5:
        results_2.append('low Activity')
    else:
        results_2.append('regular Activity')

In [47]:
# Include column in dataframe
df_merged_final['Activity_flag']=results_2

In [48]:
# Check how many low activity customers there are 
df_merged_final['Activity_flag'].value_counts(dropna=False)

regular Activity    32403157
low Activity            2884
Name: Activity_flag, dtype: int64

In [49]:
# Create subset of low activity customers
low_activity_customers= df_merged_final[df_merged_final['Activity_flag']=='low Activity']

In [50]:
# Check subset to make sure successful 
low_activity_customers.shape

(2884, 35)

In [51]:
# Export new file as pkl
low_activity_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'low_activity_customers.pkl'))

In [52]:
# Exclude sample from dataset
df_merged_final.drop(df_merged_final[df_merged_final['Activity_flag']=='low Activity'].index, inplace=True)

In [53]:
# Check to make sure successful 
df_merged_final['Activity_flag'].value_counts()

regular Activity    32403157
Name: Activity_flag, dtype: int64

In [62]:
# Viewing dataset
df_merged_final.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

   product_id product_name  aisle_id  department_id  prices  \
0         195         Soda        77              7     9.0   
1         195         Soda        77              7     9.0   
2         195         Soda        77              7     9.0   
3         195         Soda        77              7     9.0   
4         195         Soda        77              7     9.0   

     price_range_loc     busiest_day         busiest_days  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product      Least busy      Least busy days   
2  Mid-range product      Least busy      Least busy days   
3  Mid-range product      Least busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  max_order  loyalty_flag  mean_order spending_flag  \
0        Average orders         10  New customer    6.367797   Low spender   
1        Average orders         10  New customer    6.367797   Low spender   
2           Most orders         10  New customer    6.367797   Low spender   
3        Average orders         10  New customer    6.367797   Low spender   
4           Most orders         10  New customer    6.367797   Low spender   

   median_days_since_last_order          frequent_flag         Frequency_flag  \
0                     20.259259  Non-frequent customer  Non-Frequent customer   
1                     20.259259  Non-frequent customer  Non-Frequent customer   
2                     20.259259  Non-frequent customer  Non-Frequent customer   
3                     20.259259  Non-frequent customer  Non-Frequent customer   
4                     20.259259  Non-frequent customer  Non-Frequent customer   

   gender    state  age date_joined  number_of_dependants family_status  \
0  Female  Alabama   31   2/17/2019                     3       married   
1  Female  Alabama   31   2/17/2019                     3       married   
2  Female  Alabama   31   2/17/2019                     3       married   
3  Female  Alabama   31   2/17/2019                     3       married   
4  Female  Alabama   31   2/17/2019                     3       married   

   income region  total_orders     Activity_flag  
0   40423  south            59  regular Activity  
1   40423  south            59  regular Activity  
2   40423  south            59  regular Activity  
3   40423  south            59  regular Activity  
4   40423  south            59  regular Activity

In [64]:
# Create subset of regular activity customers
df_merged2= df_merged_final[df_merged_final['Activity_flag']=='regular Activity']

In [65]:
# Export new file as pkl
df_merged2.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_merged_2.pkl'))